ui/ozone/platform/wayland/wayland_buffer_manager_unittest.cc

```diff
@@ -0,0 +1,138 @@
+
+class WaylandBufferManagerTest : public testing::Test {
+ public:
+  WaylandBufferManagerTest() = default;
+  ~WaylandBufferManagerTest() override = default;
+
+ protected:
+  void SetUp() override {
+    // This expect is purely for silencing the uninteresting mock warning.
+    EXPECT_CALL(*server_.xdg_shell(),
+                UseUnstableVersion(XDG_SHELL_VERSION_CURRENT));
+
+    ASSERT_TRUE(server_.Start(XDG_SHELL_VERSION_CURRENT));
+    ASSERT_TRUE(connection_.Initialize());
+    ASSERT_TRUE(connection_.StartProcessingEvents());
+
+    wl::Object<zwp_linux_dmabuf_v1> zwp_linux_dmabuf =
+        wl::Bind<zwp_linux_dmabuf_v1>(connection_.registry_for_tests(), 1, 1);
+    manager_ = std::make_unique<WaylandBufferManager>(
+        zwp_linux_dmabuf.release(), &connection_);
+  }
+
+  WaylandBufferManager* manager() { return manager_.get(); }
+
+  base::File MakeTempFile() {
+    base::FilePath temp_path;
+    EXPECT_TRUE(base::CreateTemporaryFile(&temp_path));
+    return base::File(temp_path,
+                      base::File::FLAG_WRITE | base::File::FLAG_CREATE_ALWAYS);
+  }
+
+ private:
+  base::MessageLoopForUI message_loop_;
+  wl::FakeServer server_;
+  WaylandConnection connection_;
+
+  std::unique_ptr<WaylandBufferManager> manager_;
+
+  DISALLOW_COPY_AND_ASSIGN(WaylandBufferManagerTest);
+};
+
+TEST_F(WaylandBufferManagerTest, ValidateDataFromGpu) {
+  struct InputData {
+    bool has_file = false;
+    uint32_t width = 0;
+    uint32_t height = 0;
+    uint32_t planes_count = 0;
+    std::vector<uint32_t> strides;
+    std::vector<uint32_t> offsets;
+    std::vector<uint64_t> modifiers;
+    uint32_t format = 0;
+    uint32_t buffer_id = 0;
+  };
+
+  const uint32_t kExistingBufferId = 1;
+  const uint32_t kNonExistingBufferId = 2;
+
+  const InputData kBadInputs[] = {
+      // All zeros.
+      {},
+      // Valid file but zeros everywhereelse.
+      {true},
+      // Valid file, invalid size, zeros elsewhere.
+      {true, 1024},
+      {true, 0, 768},
+      // Valid file and size but zeros in other fields.
+      {true, 1024, 768},
+      // Vectors have different lengths.
+      {true, 1024, 768, 1, {1}, {2, 3}, {4, 5, 6}},
+      // Vectors have same lengths but strides have a zero.
+      {true, 1024, 768, 1, {0}, {2}, {6}},
+      // Vectors are valid but buffer format is not.
+      {true, 1024, 768, 1, {1}, {2}, {6}},
+      // Everything is correct but the buffer ID is zero.
+      {true, 1024, 768, 1, {1}, {2}, {6}, DRM_FORMAT_R8},
+      // Everything is correct but the buffer ID .
+      {true, 1024, 768, 1, {1}, {2}, {6}, DRM_FORMAT_R8, kExistingBufferId},
+  };
+
+  EXPECT_TRUE(manager()->CreateBuffer(MakeTempFile(), 1024, 768, {1}, {2},
+                                      DRM_FORMAT_R8, {3}, 1,
+                                      kExistingBufferId));
+  auto temp_file = MakeTempFile();
+  for (const auto& bad : kBadInputs) {
+    base::File dummy;
+    EXPECT_FALSE(manager()->ValidateDataFromGpu(
```

Comment:
That is of course possible (see the other test below) but I had decided to call the object directly because that is what the unit test should do; calling it through other layers undermines the idea of isolation which is core of unit testing.

Here I test only what the ValidateDataFromGpu method returns, cover all code paths in it, and don't want to deal with other objects and possible side effects.

I agree that we may need some integration tests but I would suggest adding them and not changing this one.